superset/assets/javascripts/explorev2/components/controls/Filter.jsx

```diff
@@ -55,7 +58,23 @@ export default class Filter extends React.Component {
     if (event && event.value) {
       value = event.value;
     }
-    this.props.changeFilter(control, value);
+    if (control === 'op') {
+      if (arrayFilterOps.indexOf(this.props.filter.op) !== -1
+        && strFilterOps.indexOf(value) !== -1) {
+        const newVal = this.props.filter.val.length > 0 ?
+          this.props.filter.val[0] : '';
+        this.props.changeFilter(['op', 'val'], [value, newVal]);
+      } else if (strFilterOps.indexOf(this.props.filter.op) !== -1
+        && arrayFilterOps.indexOf(value) !== -1) {
+        const newVal = this.props.filter.val === '' ?
+          [] : [this.props.filter.val];
+        this.props.changeFilter(['op', 'val'], [value, newVal]);
+      } else {
+        this.props.changeFilter(control, value);
+      }
+    } else {
+      this.props.changeFilter(control, value);
```

Comment:
I think `this.props.changeFilter` is called in all code paths here, maybe take it out of the outer `else` and just manage `newVal` instead in the block of code.